### PR TITLE
[CSL-1793] Allow to disable tx inputs grouping

### PR DESCRIPTION
**Pull request by Martoon-00**

### CODE REVIEW

**Comment by Martoon-00** [comment]
_Tested, description updated_

File: node/src/Pos/Client/Txp/Util.hs [hunk 1]
```diff
@@ -141,6 +145,10 @@ isCheckedTxError = \case
 -- Tx creation
 -----------------------------------------------------------------------------
 
+-- | Whether addresses can be spent entirely only.
+newtype UseGroupedInputs = UseGroupedInputs Bool
```
Comment (dcoutts) [inline node/src/Pos/Client/Txp/Util.hs]: This should be named something like an address (or input) selection policy, since that is what it is, and it should be a data type with one constructor per policy, including any policy params (so it is not an enumeration).

File: node/src/Pos/Client/Txp/Util.hs [hunk 2]
```diff
@@ -261,85 +323,97 @@ groupUtxo utxo =
...
+    => Utxo
+    -> TxOutputs
+    -> TxFee
+    -> TxCreator m TxRaw
+prepareTxRaw utxo outputs fee = do
+    UseGroupedInputs whetherGrouped <- view tcdUseGroupedInputs
+    let inputPicker =
+            if whetherGrouped
+            then groupedInputPicker
+            else plainInputPicker
```
Comment (dcoutts) [inline node/src/Pos/Client/Txp/Util.hs]: So this `if` would be replaced by a `case` on the input selection policy. This way it is naturally extendible when we add more different policy choices.

**Comment by adinapoli-iohk** [comment]
Thanks for dragging my attention over to this PR, @dcoutts .

I agree with the general sentiment, shipping a query parameter which can group & ungroup tx in a different way during tx building feels a bit ad-hoc and possibly this design doesn't scale for the general feature (i.e. "Configurable policies for tx building").

Considering we might want to provide more policies in the future (check this [article](https://medium.com/@lopp/the-challenges-of-optimizing-unspent-output-selection-a3e5d05d13ef) for lots of valuable ideas) as well as documenting them in the Swagger spec for end users (exchanges but also Daedalus programmers) I would also recommend to create a proper data type which enumerate the policies (currently just two, the "normal" one and the one for "bittrex") and allow them to be passed as JSON in the request body.

> This name is based on the mechanism not on the policy that it achieves.

Yes, we might not want to "leak" implementation details for a policy, only to give it a name for what it ultimately provides. By quoting a passage from the article I have linked:

> It’s clear that there is no “one size fits all” solution to this problem, and in fact the three broad optimization goals outlined above tend to be in direct opposition. The answer for wallet authors, given the incompatible goals, may be to give the end user more control over the wallet’s UTXO selection algorithm. It should be an advanced feature for power users, but I can imagine a simple drop down where the user selects between optimizing for “Fees / Performance / Privacy.”

In the same vein, we could have something like (I suck at names, as always):

```haskell
data UtxoGroupingPolicy =
    OptimiseForSecurity -- The general case to deal with the quantum-crypto implications
  | OptimiseForSize        -- The Bittrex case
```

We can then create a `Payment` object (which is what we'll do anyway in V1) and allow passing such policy as a optional type:

```haskell
data Payment = Payment {
  pymtUtxoGroupingPolicy :: Maybe UtxoGroupingPolicy
}
```

If `null` is passed as input we default to a sensible way (say, `OptimiseForSecurity`).

**Comment by Martoon-00** [comment]
Thanks for you comments on this!
Now grouping policy is passed in request body as either "GroupInputs" or "NoGrouping" (yeah, those unpleasant JSON strings), or just `null` which defaults to the former.

This breaks compatibility with previous version of API, since `newPayment` and `txFee` became `POST` requests accepting non-empty bodies. 
Daedalus-bridge was modified to catch these changes, and daedalus acceptance tests pass ok.

**Comment by adinapoli-iohk** [comment]
Hey @Martoon-00 , thanks for this!

> Now grouping policy is passed in request body as either "OptimizeForSecurity" or "OptimizeForSize" (yeah, those unpleasant JSON strings), or just null which defaults to the former.

I'm willing to accept this (aka passing not a valid JSON Object as input) only in the virtue of the fact we are going to overhaul the API anyway for V1 (and full disclosure, I will never approve a PR which includes a request body which is not a valid JSON _object_ 😬 ).

As long as we can communicate the Daedalus guys about the changes and they are happy with that (ping @DominikGuzei so I don't forget) I'm happy.

**Comment by Martoon-00** [comment]
Actually there was a dispute between our team and exchanges guys; in various specs we found that plain string is a valid JSON, while people from exchanges considered it not and had problems with it.
Could you shed some light on this situation?

**Comment by adinapoli-iohk** [comment]
> Could you shed some light on this thing?

Yes sure! I think it's mostly considered best practice for a couple of reasons. I have no hard evidence, but a non-comprehensive list:

- It's easier for users of the API to have _just_ JSON Objects as a response, because it means you can uniformly parse the incoming HTTP Body as a Javascript object (because as you know every json object is also a valid javascript object).

- The original RFC mentions that "A JSON text is a sequence of tokens.  The set of tokens includes six structural characters, strings, numbers, and three literal names. A JSON text is a serialized value.  Note that certain previous specifications of JSON constrained a JSON text to be an object or an array.  Implementations that generate only objects or arrays where a JSON text is called for will be interoperable in the sense that all implementations will accept these as conforming JSON texts". The TL;DR is that with JSON Objects you can never go wrong :)

- There is an old [security flaw](https://haacked.com/archive/2009/06/25/json-hijacking.aspx/) which was triggered by returning JSON arrays, but not JSON objects.

In a nutshell, I think it's just a matter of best practice & interoperability, really 😉

**Comment by Martoon-00** [comment]
Thanks, this makes things clear to me.

Okay, with last commit custom JSON-serialization options for `InputSelectionPolicy` are used, and JSON looks like
> `{ "tag": "OptimizeForSecurity" }`

Does it look better?
Daedalus still works.

**Comment by adinapoli-iohk** [comment]
> { "tag": "OptimizeForSecurity" }
> Does it look better?

It does, but as we are close to perfection, why not rename `tag` to `policy` or `groupingPolicy`? 😉  It's starting to look very good!

**Comment by Martoon-00** [comment]
Done

**Comment by adinapoli-iohk** [comment]
Thanks @Martoon-00, you are a ⭐️  !

**Comment by DominikGuzei** [comment]
Thanks for pining me @adinapoli-iohk, it's good to know since we are also overhauling our API layer and replacing the daedalus bridge with custom JS code … i sent the link to our whole team to notice the changes!

**Comment by Martoon-00** [comment]
Yep sure, I'll resolve conflicts. Will be ready till tomorrow's morning

**Comment by Martoon-00** [comment]
Appveyor fails trying to download something. I'll restart at noon in hope that it is transient issue.
